_posts/2024-08-19-quarkus-metaprogramming.adoc

```diff
@@ -0,0 +1,372 @@
+
+Here each of these entries contains an instance of the Jandex MethodInfo that carries all the necessary offline reflection information about the signature of a Java method implementing a specific REST endpoint. As anticipated we are interested in the types returned by those methods, because those are the classes of the objects that will require to be converted in json as response to the REST endpoints invocation.
+
+Now we have all the building blocks to start developing a https://quarkus.io/guides/writing-extensions#build-step-processors[BuildStep] that collects, without duplications, all those return types, represented as Jandex `ClassInfo` s, and pass them to another service that will have the responsibility of generating the bytecode of a Jackson serializer for each of them.
+
+[source, java]
+----
+@BuildStep
+public void handleEndpointParams(ResteasyReactiveResourceMethodEntriesBuildItem resourceMethodEntries, JaxRsResourceIndexBuildItem jaxRsIndex) {
+
+    	IndexView indexView = jaxRsIndex.getIndexView();
+
+    	Map<String, ClassInfo> jsonClasses = new HashMap<>();
+    	for (ResteasyReactiveResourceMethodEntriesBuildItem.Entry entry : resourceMethodEntries.getEntries()) {
+        	MethodInfo methodInfo = entry.getMethodInfo();
+        	ClassInfo returnClassInfo = indexView.getClassByName(methodInfo.returnType().name());
+        	if (returnClassInfo != null) {
+            	jsonClasses.put(returnClassInfo.name().toString(), returnClassInfo);
+        	}
+    	}
+
+    	if (!jsonClasses.isEmpty()) {
+        	// TODO: generate serializers for each returned type
+    	}
+}
+----
+
+where the JaxRsResourceIndexBuildItem provides the Jandex IndexView that allows access to all reflection information collected and indexed by Jandex. Since we already know that the json serialization will be triggered by the rest-jackson extension it is convenient to simply add this new BuildStep to the https://github.com/quarkusio/quarkus/blob/main/extensions/resteasy-reactive/rest-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/ResteasyReactiveJacksonProcessor.java[BuildStepsProcessor already present in that extension].
```

Comment:
```suggestion
where the `JaxRsResourceIndexBuildItem` provides the Jandex `IndexView` that allows access to all reflection information collected and indexed by Jandex. Since we already know that the json serialization will be triggered by the rest-jackson extension it is convenient to simply add this new `BuildStep` to the https://github.com/quarkusio/quarkus/blob/main/extensions/resteasy-reactive/rest-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/ResteasyReactiveJacksonProcessor.java[BuildStepsProcessor already present in that extension].
```